### PR TITLE
[ fix #4572 ] added piSort UnivSort funSort and Setω+n to the user-manual

### LINKED ISSUE #4572: add PiSort and UnivSort to the documentation
> `piSort` is a constructor for sorts such that if `A : s1` and `x : A |- B : s2`, then `(x : A) -> B : piSort s1 (\x -> s2)`. It should never appear in a fully typechecked program, but it is used during constraint solving when we don't know yet whether a sort is of the form `Set l`, `Prop l`, `Setω`, or `SizeUniv`. I wrote a blog post about this a while ago at https://jesper.sikanda.be/posts/agdas-new-sorts.html, maybe it should also be added to the official documentation somewhere.

_Originally posted by @jespercockx in https://github.com/agda/agda/issues/3960#issuecomment-569135666_

I found this among the very long thread #3960, but still in the current version of the user-manual not a trace of either `PiSort` and `UnivSort` exist.

Where should this be located in the documentation? Maybe Language Reference -> Sorts? If there are other observations additional to Jesper's article, please post them in this issue.

I can also remove the self-assignment if someone wants to do it.

### PR DESCRIPTION
**Pull request by pnlph**
Finally, I was editing the page of Universe levels. I think its name should also be changed maybe to `Universe hierarchy`, `Sort hierarchy` or something similar? What do you think? Since the arrival of `Setω` the levels are not the only possible universes.

### CODE REVIEW

File: doc/user-manual/language/universe-levels.lagda.rst [hunk 1]
```diff
@@ -260,6 +260,58 @@ Level) → Set n`` belongs to ``Setω`` which is not of this form. The
...
+types. The constructor ``funSort`` computes the sort of a function type
+even if the sort of the domain and the sort of the codomain are still
+unknown.
+
+Example: the sort of the function type ``∀ {A} → A → A`` with normal form
+``{A : _5} → A → A`` evaluates to ``funSort (univSort _5) (funSort _5 _5)``
+where:
+
+* ``_5`` is a metavariable that represents the sort of ``A``.
+* ``funSort _5 _5)`` is the sort of ``A → A``
```
Comment (andreasabel) [inline doc/user-manual/language/universe-levels.lagda.rst]: Extra ).

File: doc/user-manual/language/universe-levels.lagda.rst [hunk 2]
```diff
@@ -260,6 +260,58 @@ Level) → Set n`` belongs to ``Setω`` which is not of this form. The
...
+Similarly, ``PiSort s1 s2`` is a constructor that computes the sort of
+a Π-type given the sort ``s1`` of its domain and the sort ``s2`` of its
+codomain as arguments.
+
+These constructors do not represent new sorts but instead, they compute
+to the right sort once their arguments are known.
+
+More examples:
+
+* ``piSort Level (\l. Set l)`` evaluates to ``Setω``
```
Comment (andreasabel) [inline doc/user-manual/language/universe-levels.lagda.rst]: Please use `λ` and `→` in the notation for lambda-abstraction.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Found some typos.

**Comment by jespercockx** [comment]
I'd suggest this information should go on its own page about Agda's sort system, since it does not have anything to do with universe levels *per se*.

**Comment by jespercockx** [comment]
I think it would also make sense to write down the behaviour of `funSort`, `piSort`, and `univSort` on closed sorts explicitly, i.e.
```
funSort Setω     _        = Setω
funSort _        Setω     = Setω
funSort (Set a)  (Set b)  = Set (a ⊔ b)
funSort (Prop a) (Set b)  = Set (a ⊔ b)
funSort (Set a)  (Prop b) = Prop (a ⊔ b)
funSort (Prop a) (Prop b) = Prop (a ⊔ b)

piSort s1 (\x -> s2) = funSort s1 s2          (if x does not occur freely in s2)
piSort s1 (\x -> s2) = Setω                   (if x occurs freely in s2)

univSort (Set a)  = Set (lsuc a)
univSort (Prop a) = Set (lsuc a)
univSort Setω     = Setω                      (only if --omega-in-omega is enabled)
```

File: doc/user-manual/language/universe-levels.lagda.rst [hunk 3]
```diff
@@ -260,6 +260,58 @@ Level) → Set n`` belongs to ``Setω`` which is not of this form. The
...
+* ``piSort (Set l) (\_. Set l')`` evaluates to ``Set (l ⊔ l')``
+* ``univSort (Set l)`` evaluates to ``Set (lsuc l)``
+* The function type ``∀ {A} → ∀ {B} → B → A → B`` with normal form
+  ``{A : _9} {B : _7} → B → A → B`` evaluates to
+  ``piSort (univSort _9) (λ A → funSort (univSort _7)
+  (funSort _7 (funSort _9 _7)))``
+
+Note that ``funSort``, ``PiSort`` and ``UnivSort`` are not always
```
Comment (jespercockx) [inline doc/user-manual/language/universe-levels.lagda.rst]: With the current sort system, `funSort` and `piSort` are total functions on sort, but `univSort` is not (unless `--omega-in-omega` is enabled). Any uses of `univSort` will lead to a 'has bigger sort' constraint that ensures the argument is not `Setω`.

File: doc/user-manual/language/universe-levels.lagda.rst [hunk 4]
```diff
@@ -300,8 +300,8 @@ to the right sort once their arguments are known.
 
 More examples:
 
-* ``piSort Level (\l. Set l)`` evaluates to ``Setω``
-* ``piSort (Set l) (\_. Set l')`` evaluates to ``Set (l ⊔ l')``
+* ``piSort Level (λ → Set l)`` evaluates to ``Setω``
+* ``piSort (Set l) (λ → Set l')`` evaluates to ``Set (l ⊔ l')``
```
Comment (gallais) [inline doc/user-manual/language/universe-levels.lagda.rst]: Looks like we are missing a bound `l` in the first case and an underscore in the second.

File: doc/user-manual/language/universe-levels.lagda.rst (hunk 3 above)
Comment (pnlph) [reply doc/user-manual/language/universe-levels.lagda.rst]: OK, I changed the paragraph, please take a look.

File: doc/user-manual/language/universe-levels.lagda.rst (hunk 4 above)
Comment (pnlph) [reply doc/user-manual/language/universe-levels.lagda.rst]: Thanks!

**Comment by pnlph** [comment]
I added a new `sort-system` page to the documentation, like Jesper proposed. I added it also to the index of the language folder, I hope this suffices...

There are code bits describing the behaviour as proposed, introduced by `.. code-block::` but without `agda`. I hope they are not being typechecked. Is there another way of introducing dummy code?

File: doc/user-manual/language/sort-system.lagda.rst [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+funSort
+-------
+
+Agda’s internal representation of sorts implements a constructor (sort
+metavariable) representing an unknown sort. The constraint solver can
+compute these sort metavariables, just like it does when computing
+regular term metavariables.
+
+But this sort metavariable need other constructors to solve function
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst]: need --> needs

The sentence is a bit unclear.

I think there should be a basic explanation that if in function type `A -> B` domain `A` has sort `sA` and codomain `B` has sort `sB`, then the function type has sort `funSort sA sB`. If `sA` and `sB` are known, `funSort sA sB` can be computed to a sort value.


Like there should be a basic explanation that if in dependent function type `(x : A) -> B` domain `A` has sort `sA` and codomain `B` has sort `sB` dependent on `x : A`, then the function type has sort `piSort sA (\ x -> sB)`.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks.  Looks good in general, I have one request, though.

File: doc/user-manual/language/sort-system.lagda.rst (hunk 5 above)
Comment (pnlph) [reply doc/user-manual/language/sort-system.lagda.rst]: > need --> needs

Thanks, this happens when committing at night...

> I think there should be a basic explanation...

I agree, please see how it reads now.

File: doc/user-manual/language/sort-system.lagda.rst [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+Similarly, ``PiSort s1 s2`` is a constructor that computes the sort of
+a Π-type given the sort ``s1`` of its domain and the sort ``s2`` of its
+codomain as arguments.
+
+To understand how ``PiSort`` works in general, we set the following scenario:
+
+* ``sA`` and ``sB`` are two (possibly different) sorts.
+* ``A : sA``, meaning that ``A`` is a type that has sort ``sA``.
+* ``x : A``, meaning that ``x`` has type ``A``.
+* ``B : sB``, meaning that B is a type (possibly different than ``A``) that
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst]: B --> ``B``  sB --> ``sB``

File: doc/user-manual/language/sort-system.lagda.rst [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  has sort sB.
+
+Under these conditions, we can build the dependent function type
+``(x : A) → B : PiSort sA (λ x → sB)``. This type signature means that the
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst]: PiSort --> piSort  (several times)

File: doc/user-manual/language/sort-system.lagda.rst [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+* ``piSort Level (λ l → Set l)`` evaluates to ``Setω``
+* ``piSort (Set l) (λ _ → Set l')`` evaluates to ``Set (l ⊔ l')``
+* ``univSort (Set l)`` evaluates to ``Set (lsuc l)``
+* The function type ``∀ {A} → ∀ {B} → B → A → B`` with normal form
+  ``{A : _9} {B : _7} → B → A → B`` evaluates to
+  ``piSort (univSort _9) (λ A → funSort (univSort _7)
+  (funSort _7 (funSort _9 _7)))``
+
+Note that ``funSort`` and ``PiSort`` are total functions on sort. But
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst]: PiSort --> piSort

File: doc/user-manual/language/sort-system.lagda.rst [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+``UnivSort`` is not always well-defined. Eg. without adding the
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst]: UnivSort --> univSort

File: doc/user-manual/language/sort-system.lagda.rst [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+funSort
+-------
+
+Agda’s internal representation of sorts implements a constructor (sort
+metavariable) representing an unknown sort. The constraint solver can
+compute these sort metavariables, just like it does when computing
+regular term metavariables.
+
+But this sort metavariable needs other constructors to solve function
+types. The constructor ``funSort`` computes the sort of a function type
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst:47]: It should be made very clear that the sort constructors `funSort` `univSort` `piSort` are internal constructors that are sometimes printed, but can not be entered by the user.

**Review by andreasabel — CHANGES_REQUESTED** [review]
Getting very good.
- Some clarification still need.
- Some case issues.

File: doc/user-manual/language/sort-system.lagda.rst (hunk 10 above)
Comment (pnlph) [reply doc/user-manual/language/sort-system.lagda.rst:47]: I added a note before the explanations.

**Comment by jespercockx** [comment]
I just merged #4595 so now the documentation should also mention the new sorts `Setωᵢ`. This also means the `Has bigger sort` constraint is now no longer necessary and has been removed. Do you want to update the documentation in this PR or should I do it directly?

**Comment by pnlph** [comment]
I will see how the changelog looks like, and I will update here.

File: doc/user-manual/language/sort-system.lagda.rst [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+Any uses of ``univSort`` will lead to a 'has bigger sort' constraint that
+ensures the argument is not ``Setω``.
+
+.. _set-omega-plus-n:
+
+Sorts ``Setωᵢ``
+---------------
+
+Agda implements sorts of the form ``Setωᵢ``, where ``i`` is a natural number
+``i : Nat``. These sorts constitute a second hierarchy ``Setωᵢ : Setωᵢ₊₁``
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst]: Is really `i : Nat`?  This looks like `i` can be any Agda-expression of type `Nat` (as in `BUILTIN NATURAL Nat`).  Isn't it rather that `i` is a numeral in decimal representation, i.e. `0`, `1`, ... etc.?

File: doc/user-manual/language/sort-system.lagda.rst [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+the one implemented in the standard hierarchy ``Setᵢ : Setᵢ₊₁`` that we
+introduced in the section :ref:`Universe Levels <universe-levels>`.
+
+But, unlike the standard hierarchy of universes ``Setᵢ``, this second
+hierarchy ``Setωᵢ`` does not support universe polymorphism. This means that
+it is not possible to quantify over *all* Setω+n at once. For example, the
+expression ``∀ {i} (A : Setω i) → A → A`` would not be a well-formed agda
+term.
+
+You can also refer to these sorts with the alternative syntax ``Setωi``.
```
Comment (andreasabel) [inline doc/user-manual/language/sort-system.lagda.rst:196]: Is `Setω0` also legal (and the one where `0` is a subscript)?

**Review by andreasabel — CHANGES_REQUESTED** [review]
Thanks for fixing my issues and for the new material on Set(omega)i.

As you can see, I have also issues with the new material. 😉 

For the sake of finally merging this PR, may I suggest to only fix issues and not add further new material?  New material (that is not strictly related) can go into a new PR.  Otherwise, I am afraid we will play this ping-pong game between contributor and reviewer for a long time... 😉 

UPDATE: I saw @jespercockx requested the new material.  Oh well, you are in a double bind between two bosses pulling into different directions. 🤣

**Comment by andreasabel** [comment]
Note: the `user-manual-test` fails with this PR: https://travis-ci.org/github/agda/agda/jobs/690102891
```
The name of the top level module does not match the file name. 
The module sort.system should be defined in one of the following files: ...
```

File: doc/user-manual/language/sort-system.lagda.rst (hunk 12 above)
Comment (pnlph) [reply doc/user-manual/language/sort-system.lagda.rst:196]: I would say so, but let's ask @jespercockx to double-check: are `Setω0` and `Setω₀` legal keywords after #4609 was merged? They behave like their counterparts `Set0` and `Set₀` , right?

Comment (jespercockx) [reply doc/user-manual/language/sort-system.lagda.rst:196]: They are legal but they are not keywords, you can rename `Setω` to anything such as `BleeBlorkl` in which case you can write `BleeBlorkl0` or `BleeBlorkl₀` and they mean the same thing as `BleeBlorkl` (which is to say, `Setω`).

**Comment by pnlph** [comment]
> This is not correct, the `i` can _only_ be a natural number literal, not an arbitrary expression of type `Nat` (how would that even work?)

OK, I thought that `Setω+n` would also work with arbitrary expressions of type `Nat` that may contain free variables _via_ `funSort`, `univSort` and `piSort`... or different constructors `funSortω`, `univSortω` and `piSortω` must be implemented for this to work?

File: doc/user-manual/language/sort-system.lagda.rst (hunk 12 above)
Comment (pnlph) [reply doc/user-manual/language/sort-system.lagda.rst:196]: OK, then if `Set`, `Setω` and all its derivations are not keywords, I have to remove them from the list of keywords in [lexical structure](https://agda.readthedocs.io/en/latest/language/lexical-structure.html), right?

**Comment by jespercockx** [comment]
> 
> 
> > This is not correct, the `i` can _only_ be a natural number literal, not an arbitrary expression of type `Nat` (how would that even work?)
> 
> OK, I thought that `Setω+n` would also work with arbitrary expressions of type `Nat` that may contain free variables _via_ `funSort`, `univSort` and `piSort`... or different constructors `funSortω`, `univSortω` and `piSortω` must be implemented for this to work?

No, the index of `Setωi` cannot refer to any free variables, it is really just a literal natural number. I'm not sure why you would need `funSortω`, `univSortω` and `piSortω`, the normal `funSort`, `univSort`, and `piSort` still work for `Setωi`. For example, `piSort s (\x -> Setωi)` will always reduce to `funSort s Setωi` since there is no way for the `i` to depend on the variable `x`.

To put it in other words: the fact that `Setωi` cannot depend on any variables is an intentional design decision. Allowing that would necessitate an additional universe `Set2ω`, which would then soon lead to `Set2ω+1`, etc, which we'd like to avoid.

File: doc/user-manual/language/sort-system.lagda.rst (hunk 12 above)
Comment (jespercockx) [reply doc/user-manual/language/sort-system.lagda.rst:196]: Ah yes, I forgot to update that list in my PR. `Set` and `Prop` should indeed be removed from that list, though you don't necessarily have to add it to this PR (as @andreasabel said, it is better to keep PR's focused on one thing).

Comment (pnlph) [reply doc/user-manual/language/sort-system.lagda.rst:196]: OK perfect, i reverted the inclusion of `Setω` as keyword, you can do it for `Set` in yours.

**Review by andreasabel — APPROVED** [review]
Looks good now.

**Comment by andreasabel** [comment]
Thanks @pnlph !

**Pull request merged**